File: 프로그래머스/1/42862. 체육복/체육복.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <set>

using namespace std;

int solution(int n, vector<int> lost, vector<int> reserve) {
    
    set<int> lost_set(lost.begin(), lost.end());
    set<int> reserve_set(reserve.begin(), reserve.end());
    vector<int> remove;
    
    for(auto lost: lost_set) {
        if (reserve_set.count(lost)) {
            remove.push_back(lost);
        }
    }
    
    for(auto r: remove) {
        lost_set.erase(r);
        reserve_set.erase(r);
    }
    
    int answer = n - lost_set.size();
    
    for(auto x: lost_set) {
        if (reserve_set.count(x-1)) {
            reserve_set.erase(x-1);
            answer++;
        } else if (reserve_set.count(x+1)) {
            reserve_set.erase(x+1);
            answer++;
        }
    }
    
    return answer;
}

File: 프로그래머스/1/42862. 체육복/체육복.swift
func solution(_ n:Int, _ lost:[Int], _ reserve:[Int]) -> Int {
    var lostSet = Set(lost)
    var reserveSet = Set(reserve)
    
    let both = lostSet.intersection(reserveSet)
    lostSet.subtract(both)
    reserveSet.subtract(both)
    
    var answer = n - lostSet.count
    
    for l in lostSet.sorted() {
        if reserveSet.contains(l-1) {
            answer += 1
            reserveSet.remove(l-1)
        } else if reserveSet.contains(l+1) {
            answer += 1
            reserveSet.remove(l+1)
        }
    }
    return answer
}